### PR TITLE
first draft

### PR DESCRIPTION
**Pull request by kfirbio**

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+name = "geodb"
+version = "0.1.0"
+description = ""
+authors = ["Kfir"]
+readme = "README.md"
+
+[tool.poetry.dependencies]
+python = "^3.11"
+fastapi = "^0.92.0"
+uvicorn = {extras = ["standard"], version = "^0.20.0"}
```
Comment (asafalon1) [inline pyproject.toml:11]: why the extras=standard? what does that do?

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ipaddress = "^1.0.23"
+pytest = "^7.2.1"
+flake8 = "^6.0.0"
+isort = "^5.12.0"
+mypy = "^1.0.1"
```
Comment (asafalon1) [inline pyproject.toml:16]: you dont want to install pytest or mypy when you deploy your service, only when you run and test it locally. (hint - poetry supports the feature you are looking for ;) )

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+
+[build-system]
+requires = ["poetry-core"]
+build-backend = "poetry.core.masonry.api"
```
Comment (asafalon1) [inline pyproject.toml:21]: do you need this? what does this do?

File: scripts/lint.sh [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
+# run various linters
+set -e
+echo "running isort..."
+python -m isort . -c
```
Comment (asafalon1) [inline scripts/lint.sh:5]: isort should also use max lines of 120 characters like flake8, otherwise they will fail each other.

File: scripts/lint.sh [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+echo "running flake8..."
+python -m flake8 --max-line-length 120
+echo "running mypy..."
+python3 -m mypy --explicit-package-bases --show-error-codes GeoDB
```
Comment (asafalon1) [inline scripts/lint.sh:9]: why --explicit-package-bases? or --show-error-codes? what do these options do? and why do you need them?

Comment (asafalon1) [inline scripts/lint.sh:9]: seems like you do a lot of configuring using CLI args, is there another method to do this?

File: tests/unittests/geo.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        "min": ipaddress.ip_address("128.0.0.0"),
+        "max": ipaddress.ip_address("142.1.0.0"),
+        "lat": -33.8,
+        "long": 151.2,
+        "city": "sidney2"
+    }]
+    geo_data = GeoDataLib(mock_db)
+
+    res = geo_data.calc_geo_data("128.0.0.0")
+    assert res.city == "sidney2"
```
Comment (asafalon1) [inline tests/unittests/geo.py:25]: add more test cases - the max ip in the range, one in between etc.

File: tests/unittests/geo.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+
+def test_get_geo_not_found():
+    mock_db = []
+    geo_data = GeoDataLib(mock_db)
+    with pytest.raises(ValueError):
+        geo_data.calc_geo_data("121.21.21.21")
```
Comment (asafalon1) [inline tests/unittests/geo.py:32]: testing you get not found when the DB is empty is an edge case, you forgot to test the real case - out of range ip.

File: tests/unittests/geo.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+import pytest
+
+from GeoDB.core.geo_data_lib import GeoDataLib
+
+
+def test_get_geo_invalid():
+    geo_data = GeoDataLib()
+    with pytest.raises(ValueError):
+        geo_data.calc_geo_data("no valid")
```
Comment (asafalon1) [inline tests/unittests/geo.py:11]: test more invalid cases?

**Review by asafalon1 — CHANGES_REQUESTED** [review]

File: .github/workflows/test.yml [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        python -m pip install --upgrade pip wheel
+        pip install poetry
+
+    - name: Setup Cache
+      uses: actions/cache@v2
+      with:
+        path: ~/.cache/pypoetry
+        key: ${{ runner.os }}-pip-
+        restore-keys: |
+          ${{ runner.os }}-pip-
```
[the remark this answers is not in the document]
Comment (kfirbio) [reply .github/workflows/test.yml:35]: dependencies caching, it should saves the time for fetching the dependencies again and again each build

File: GeoDB/main.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+    def __int__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+
+app = GeoDBApp()
+
+
+@app.get("/health")
+def read_root():
```
[the remark this answers is not in the document]
Comment (kfirbio) [reply GeoDB/main.py:17]: opps that was the default function on the fastapi template, changed to get_health_status

File: GeoDB/main.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
+from fastapi import FastAPI, HTTPException, Path
+
+from GeoDB.core.geo_data_lib import GeoDataLib
+
+
+class GeoDBApp(FastAPI):
+    GeoDataLibInstance = GeoDataLib()
```
[the remark this answers is not in the document]
Comment (kfirbio) [reply GeoDB/main.py:7]: done

File: GeoDB/main.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    return {"Status": "OK"}
+
+
+@app.get("/geo-data/{ip}")
+def get_geo_data(ip: str = Path(
+    regex=r"^(?:(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])\.){3}"
+          "(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])$")):
+    try:
+        res = app.GeoDataLibInstance.calc_geo_data(ip)
+        return res
```
[the remark this answers is not in the document]
Comment (kfirbio) [reply GeoDB/main.py:27]: done

File: GeoDB/core/geo_data_lib.py [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+            "lat": -49,
+            "long": -123.4
+        }
+    ]
+
+    def __init__(self, db=None):
+        if db is not None:
+            self._db = db
+
+    def calc_geo_data(self, input: str) -> GeoDataResponse:
```
[the remark this answers is not in the document]
Comment (kfirbio) [reply GeoDB/core/geo_data_lib.py:35]: done

File: GeoDB/core/geo_data_lib.py [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+
+class GeoDataLib:
+    _db: List[Any] = [
+        {
+            "min": ipaddress.ip_address("128.0.0.0"),
+            "max": ipaddress.ip_address("142.1.0.0"),
+            "lat": -33.8,
+            "long": 151.2,
+            "city": "sidney"
+        },
```
[the remark this answers is not in the document]
Comment (kfirbio) [reply GeoDB/core/geo_data_lib.py:15]: done

File: GeoDB/core/geo_data_lib.py [hunk 15]
```diff
@@ -0,0 +1,40 @@
+import ipaddress
+from typing import Any, List
+
+from GeoDB.schemas.api import GeoDataResponse
+
+
+class GeoDataLib:
+    _db: List[Any] = [
```
[the remark this answers is not in the document]
Comment (kfirbio) [reply GeoDB/core/geo_data_lib.py:8]: done

File: GeoDB/schemas/api.py [hunk 16]
```diff
@@ -0,0 +1,9 @@
+class GeoDataResponse:
+    lat: str
+    lng: str
+    city: str
+
+    def __init__(self, lat: str, lng: str, city: str):
+        self.lat = lat
+        self.lng = lng
+        self.city = city
```
[the remark this answers is not in the document]
Comment (kfirbio) [reply GeoDB/schemas/api.py:9]: done

File: pyproject.toml (hunk 1 above)
Comment (kfirbio) [reply pyproject.toml:11]: more dependencies which not actually used, removing

File: tests/unittests/geo.py (hunk 6 above)
Comment (kfirbio) [reply tests/unittests/geo.py]: done

File: scripts/lint.sh (hunk 5 above)
Comment (kfirbio) [reply scripts/lint.sh:9]: avoid placing __init__ anywhere

Comment (kfirbio) [reply scripts/lint.sh]: moved to setup.cfg

File: Dockerfile [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+COPY ./poetry.lock /tmp/poetry.lock
+COPY ./pyproject.toml /tmp/pyproject.toml
+RUN cd /tmp && poetry install --no-dev --no-root
+
+COPY ./GeoDB /GeoDB/GeoDB
+COPY ./pyproject.toml /GeoDB/pyproject.toml
+WORKDIR /GeoDB/
+Expose 8000
+ENV PYTHONPATH=/GeoDB
+ENV PYTHONOPTIMIZE=1
```
[the remark this answers is not in the document]
Comment (kfirbio) [reply Dockerfile:17]: PYTHONOPTIMIZE - optimizing for prod (debug off. no asserts)
PYTHONPATH - python lib locations, setting it to the main app path

File: Dockerfile [hunk 18]
```diff
@@ -0,0 +1,19 @@
+FROM python:3.11
+
+RUN pip install poetry
+RUN poetry config virtualenvs.create false
+WORKDIR /GeoDB/
+
+
+COPY ./poetry.lock /tmp/poetry.lock
+COPY ./pyproject.toml /tmp/pyproject.toml
+RUN cd /tmp && poetry install --no-dev --no-root
```
[the remark this answers is not in the document]
Comment (kfirbio) [reply Dockerfile:8]: --no-dev for removing dev related libs (tests,lints) 
--no-root is for avoiding packing the root package as dependency

File: .github/workflows/test.yml [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+    - name: Install dependencies
+      run: |
+        poetry install --no-root
```
[the remark this answers is not in the document]
Comment (kfirbio) [reply .github/workflows/test.yml:38]: to avoid packing the main package as dependency (same as dockerfile)

File: pyproject.toml (hunk 3 above)
Comment (kfirbio) [reply pyproject.toml:24]: this is for fixing poetry to manage python build. can be alter to different manager but not totally removed

File: pyproject.toml [hunk 20]
```diff
@@ -0,0 +1,25 @@
...
+authors = ["Kfir"]
+readme = "README.md"
+
+[tool.poetry.dependencies]
+python = "^3.11"
+fastapi = "^0.92.0"
+ipaddress = "^1.0.23"
+pytest-asyncio = "^0.20.3"
+uvicorn = "^0.20.0"
+yellowbox = "^0.8.2"
```
Comment (asafalon1) [inline pyproject.toml:14]: yellowbox should be a dev dependency

File: pyproject.toml [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+version = "0.1.0"
+description = ""
+authors = ["Kfir"]
+readme = "README.md"
+
+[tool.poetry.dependencies]
+python = "^3.11"
+fastapi = "^0.92.0"
+ipaddress = "^1.0.23"
+pytest-asyncio = "^0.20.3"
```
Comment (asafalon1) [inline pyproject.toml:12]: pytest should be a dev dependency

File: pyproject.toml (hunk 3 above)
Comment (asafalon1) [reply pyproject.toml:26]: I just removed it from one of my services and it works fine, are you sure you cant remove it?

File: tests/blackox/test_app.py [hunk 22]
```diff
@@ -0,0 +1,36 @@
+import pytest
```
Comment (asafalon1) [inline tests/blackox/test_app.py:1]: folder name is blackox 😆

File: tests/blackox/test_app.py [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+from async_asgi_testclient import TestClient
+
+
+@pytest.mark.asyncio
+async def test_health():
+    from GeoDB.main import app
+    async with TestClient(app) as client:
```
Comment (asafalon1) [inline tests/blackox/test_app.py:8]: why dont you make a single fixture for the test client that has a higher scope?
if you had a thousand tests it will take forever to run if you create the client every time.

File: tests/unittests/geo.py [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+    with pytest.raises(ValueError):
+        geo_data.calc_geo_data("128.0.0.0aa")
+    with pytest.raises(ValueError):
+        geo_data.calc_geo_data("999.999.999.999")
```
Comment (asafalon1) [inline tests/unittests/geo.py:16]: pytest lets you do this more easily with a decorator called parameterize. look it up.

**Review by asafalon1 — CHANGES_REQUESTED** [review]

**Review by asafalon1 — APPROVED** [review]
